Fix flash wear leveling sector calculation

Diff-fusion opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

Currently the calculation of the last usable flash sector doesn't work properly if `last_sector` starts beyond the end of the actual flash size:
https://github.com/qmk/qmk_firmware/blob/172c3496756fb5468c778c2536b4334f25eb883b/platforms/chibios/drivers/wear_leveling/wear_leveling_efl.c#L87-L92
Here the `last_sector` is calculated from the `first_sector` but also `i` is included in the calculation. But `i` is incremented in each loop iteration and thus an increasing offset is added.
The solution here is to first calculate the number of the last usable sector and then use another loop to calculate the first sector.

<!--- Describe your changes in detail here. -->

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [x] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

sigprof commented on platforms/chibios/drivers/wear_leveling/wear_leveling_efl.c:
```diff
@@ -84,12 +84,14 @@ bool backing_store_init(void) {
 
     // Work out how many sectors we want to use, working backwards from the end of the flash
     flash_sector_t last_sector = desc->sectors_count - WEAR_LEVELING_EFL_OMIT_LAST_SECTOR_COUNT;
+
+    // skip sectors that are past the actual flash size
+    while (flashGetSectorOffset(flash, last_sector) >= flash_size) {
+        last_sector--;
+    }
```
Applying `WEAR_LEVELING_EFL_OMIT_LAST_SECTOR_COUNT` first looks wrong (probably whoever added this code assumed that `desc->sectors_count` should be correct, but this might not be the case).  There are no real users of that option in the repo though.

sigprof commented on platforms/chibios/drivers/wear_leveling/wear_leveling_efl.c:
```diff
@@ -84,12 +84,14 @@ bool backing_store_init(void) {
     flash_sector_t last_sector = desc->sectors_count - WEAR_LEVELING_EFL_OMIT_LAST_SECTOR_COUNT;
+
+    // skip sectors that are past the actual flash size
+    while (flashGetSectorOffset(flash, last_sector) >= flash_size) {
```
This code would trigger undefined behavior on the first iteration if EFL uses the `desc->sectors` array (e.g., on STM32F4xx), because the valid range of sector numbers is [0…`desc->sectors_count-1`].  The original code (before it had been broken by the change which added `WEAR_LEVELING_EFL_OMIT_LAST_SECTOR_COUNT`) did not have this problem, because it subtracted at least 1 from `desc->sectors_count`.

Also should we guard against more pathological cases, like no free sectors, or even a sector being partially out of range of the reported flash size?  Although with the current error handling the only option is apparently to halt with a message that nobody would see.

sigprof reviewed (changes requested):
Looks like this code won't work on STM32F4xx (and any other chips with non-uniform flash sectors), because `flashGetSectorOffset(flash, desc->sectors_count)` would read beyond the end of the `desc->sectors` array.  So the code which finds the actual number of available flash sectors needs to be careful to avoid passing invalid sector numbers to ChibiOS functions.

Diff-fusion commented:
Thanks for the review. I fixed the issues you mentioned and added some additional checks.
One remaining issue is that flash sectors that are used for the firmware code could be used by the wear_leveling driver. I don't see any checks to prevent that. But I'm not that experienced with QMK and don't know what would be the best way to do this.
Error handling is a bit problematic. Exiting with `chSysHalt` is better than just crashing. But maybe compile time checks could be used to prevent some of the error cases.

Diff-fusion commented:
Hey, is there anything missing to get this merged?

tzarc commented:
> Hey, is there anything missing to get this merged?

From you, nothing at this stage. Needs two approving reviews for merge; we'll get to it when we can.
